pkg/controller/bluegreendeployment/helpers.go

```diff
@@ -0,0 +1,302 @@
+	blueRS, err := createReplicaSetIfNotFound(b, c, BlueRSName, BlueColor)
+	if err != nil {
+		return nil, nil, fmt.Errorf("when creating blue ReplicaSet, %v", err)
+	}
+
+	greenRS, err := createReplicaSetIfNotFound(b, c, GreenRSName, GreenColor)
+	if err != nil {
+		return nil, nil, fmt.Errorf("when creating green ReplicaSet, %v", err)
+	}
+
+	return blueRS, greenRS, nil
+}
+
+// createReplicaSetsIfNotFound creates a ReplicaSet with given color label if it is not found.
+func createReplicaSetIfNotFound(b *v1alpha1.BlueGreenDeployment, c *kubernetes.Clientset, rsName, rsColor string) (*appsv1.ReplicaSet, error) {
+	ns := b.Namespace
+	rs, err := c.AppsV1().ReplicaSets(ns).Get(rsName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			rs, err = c.AppsV1().ReplicaSets(ns).Create(newReplicaSet(b, rsColor))
+			if err != nil {
+				return nil, fmt.Errorf("failed to create ReplicaSet %q: %v", rsName, err)
+			}
+			if err = waitAllReplicaSetPodsAvailable(c, rs); err != nil {
+				// When not all pods become available within timeout, it may be due to invalid
+				// container field input(s) such as invalid image name. We should not return an
+				// error to ensure the ReplicaSet to be made inactive with valid container fields
+				// is scaled down later down the sync call tree. Logging the failure as a message
+				// should suffice as this controller does not support auto-rollback feature.
+				log.Printf("some pods for ReplicaSet %q failed to become available: %v", rs.Name, err)
+			}
+			return rs, nil
+		} else {
+			return nil, fmt.Errorf("failed to get ReplicaSet %q: %v", rsName, err)
+		}
+	}
+
+	return rs, nil
+}
+
+// updateBlueGreenDeployment updates a BlueGreenDeployment object based on given update function via polling.
+func updateBlueGreenDeployment(c controllerv1alpha1client.ControllerV1alpha1Interface, bgdName, ns string, updateFunc func(*v1alpha1.BlueGreenDeployment)) (*v1alpha1.BlueGreenDeployment, error) {
+	var blueGreenDeployment *v1alpha1.BlueGreenDeployment
+	bgdClient := c.BlueGreenDeployments(ns)
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		bgd, err := bgdClient.Get(bgdName, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get BlueGreenDeployment object: %v", err)
+		}
+		updateFunc(bgd)
+		blueGreenDeployment, err = bgdClient.Update(bgd)
+		if err != nil {
+			return fmt.Errorf("failed to update BlueGreenDeployment object: %v", err)
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return blueGreenDeployment, nil
+}
+
+// waitAllReplicaSetPodsAvailable waits all ReplicaSet pods to become available.
+func waitAllReplicaSetPodsAvailable(c *kubernetes.Clientset, replicaSet *appsv1.ReplicaSet) error {
+	desiredGeneration := replicaSet.Generation
+	return wait.PollImmediate(PollInterval, PollTimeout, func() (bool, error) {
```

Comment:
When we last talked I thought the plan was to do background deletion.

Reply:
Yes, but this is another function to wait all ReplicaSet pods to become available.